clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,239 +23,410 @@ package common
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.start(config.ClientSet)
+	case "gather":
+		timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.gather(config.ClientSet, timeout)
+	case "stop":
+		w.stop()
+		return summaries, nil
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
-	if err != nil {
-		return summaries, err
+}
+
+func (w *waitForControlledPodsRunningMeasurement) start(c clientset.Interface) error {
+	if w.informer != nil {
+		glog.Infof("%v: wait for controlled pods measurement already running", waitForControlledPodsRunningName)
+		return nil
 	}
-	fieldSelector, err := util.GetStringOrDefault(config.Params, "fieldSelector", "")
-	if err != nil {
-		return summaries, err
+	glog.Infof("%v: starting wait for controlled pods measurement...", waitForControlledPodsRunningName)
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = w.fieldSelector
+		options.LabelSelector = w.labelSelector
 	}
-	timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), w.kind+"s", w.namespace, optionsModifier)
+	obj, err := runtimeobjects.GetObjectForKind(w.kind)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+	w.isRunning = true
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, obj, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addF := func() {
+				runtimeObj, ok := obj.(runtime.Object)
+				if !ok {
+					return
+				}
+				if err := w.updateOpResourceVersion(runtimeObj); err != nil {
+					return
+				}
+				w.handleObject(c, runtimeObj)
+			}
+			w.queue.Add(&addF)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			updateF := func() {
+				oldRuntimeObj, ok := oldObj.(runtime.Object)
+				if !ok {
+					return
+				}
+				newRuntimeObj, ok := newObj.(runtime.Object)
+				if !ok {
+					return
+				}
+				// Increment version regardless of change in spec.
+				if err := w.updateOpResourceVersion(newRuntimeObj); err != nil {
+					return
+				}
+				if isEqual, err := runtimeobjects.IsEqualRuntimeObjectsSpec(oldRuntimeObj, newRuntimeObj); err != nil || isEqual {
+					return
+				}
+				w.handleObject(c, newRuntimeObj)
+			}
+			w.queue.Add(&updateF)
+		},
+		DeleteFunc: func(obj interface{}) {
+			deleteF := func() {
+				runtimeObj, ok := obj.(runtime.Object)
+				if !ok {
+					return
+				}
+				if err := w.updateOpResourceVersion(runtimeObj); err != nil {
+					return
+				}
+				w.deleteObject(runtimeObj)
+			}
+			w.queue.Add(&deleteF)
+		},
+	})
+
+	for i := 0; i < waitForControlledPodsWorkers; i++ {
+		w.wg.Start(func() {
+			for {
```

Comment:
Opaque this into a separate worker function.

Reply:
Done.